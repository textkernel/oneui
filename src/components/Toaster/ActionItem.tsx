import * as React from 'react';
import { ENTER_KEY } from '../../constants';
import styles from './Toaster.scss';
import { bem } from '../../utils';

export interface ActionProps {
    /** Action object, containing text, callback and/or href to action  */
    text: string;
    callback?: () => void;
    href?: string;
}

const { elem } = bem('Toast', styles);

export const ActionItem = (props: { action: ActionProps; dismissToast: () => void }) => {
    const { action, dismissToast } = props;
    if (action.callback) {
        const handleOnClick = () => {
            action.callback!();
            dismissToast();
        };
        const handleKeyDown = (e) => {
            if (e.key === ENTER_KEY) {
                handleOnClick();
            }
        };
        return (
            <div
                onClick={handleOnClick}
                onKeyDown={handleKeyDown}
                {...elem('toastAction')}
                tabIndex="0"
                role="button"
            >
                {action.text}
            </div>
        );
    }
    if (action.href) {
        return (
            <a
                onClick={dismissToast}
                href={action.href}
                target="_blank"
                rel="noreferrer"
                {...elem('toastAction')}
            >
                {action.text}
            </a>
        );
    }
    return null;
};

ActionItem.displayName = 'ActionItem';
