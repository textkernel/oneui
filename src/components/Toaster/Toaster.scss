@use 'sass:map';
@import '../../themes/oneui/oneui';

$contexts: (info, success, cautious, critical);

.Toast {
    display: flex;
    flex-direction: row;
    gap: var(--space-150);
    min-width: 400px;
    padding: var(--space-200);
    border: 1px solid;
    border-radius: var(--space-150);

    &__contextIcon {
        padding: 0;
        margin: 0;
        font-size: var(--space-300);
        @each $context in $contexts {
            &--context_#{$context} {
                fill: var(--color-icon-#{$context}-default);
            }
        }
    }

    &__content {
        flex-grow: 1;
        display: flex;
        flex-direction: column;
        gap: var(--space-150);
    }

    &__message {
        gap: var(--space-100);
        display: flex;
        flex-direction: column;
        margin: 0;
        font-family: var(--font-family-primary);
        h3,
        p {
            margin: 0;
        }
    }

    &__actions {
        display: flex;
        flex-direction: row;
        align-items: center;
        gap: var(--space-100);
        color: var(--color-link-default);
    }

    &__toastAction {
        composes: OneUI-label-text-bold from global;
        text-decoration: none;
        color: var(--color-link-default);
        cursor: pointer;
        padding: 0;
        margin: 0;
    }

    &__closeButton {
        display: flex;
        align-items: flex-start;
        height: var(--space-300);
        width: var(--space-300);
        padding: 0;
        margin: 0;
        cursor: pointer;
        background-color: var(--transparent);
        border: none;
        outline: 0;

        &:focus {
            box-shadow: 0px 0px 0px 2px var(--color-background-neutral-subtlest-pressed);
            border-radius: var(--space-50);
        }
    }

    &__closeIcon {
        font-size: var(--space-300);
        color: var(--color-icon-subtle);
        fill: var(--color-icon-subtle);
    }

    @each $context in $contexts {
        &--context_#{$context} {
            background-color: var(--color-background-#{$context}-subtlest-default);
            border-color: var(--color-border-#{$context}-subtle-#{$context});
        }
    }
}

.Toaster {
    --width: unset;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: flex-end;
    z-index: 1000;
    &[data-y-position='bottom'] {
        bottom: var(--space-200);
    }
}
