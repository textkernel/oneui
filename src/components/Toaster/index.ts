export { toast, ToastProps, Toaster } from './Toaster';
