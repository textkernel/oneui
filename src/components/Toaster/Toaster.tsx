/* eslint-disable import/no-unresolved */
import * as React from 'react';
import { Toaster as SonnerToaster, toast as sonnerToast } from 'sonner';
import CloseIcon from '@material-design-icons/svg/round/close.svg';
import CautiousIcon from '@material-design-icons/svg/round/warning.svg';
import CriticalIcon from '@material-design-icons/svg/round/error.svg';
import InfoIcon from '@material-design-icons/svg/round/info.svg';
import SuccessIcon from '@material-design-icons/svg/round/check_circle.svg';
import { assertNever } from '../../utils/misc';
import styles from './Toaster.scss';
import { Context } from '../../constants';
import { bem } from '../../utils';
import { ActionItem, ActionProps } from './ActionItem';

export interface ToastProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Short and descriptive title */
    title?: string;
    /** Body content */
    description: string;
    /** The Toast context (e.g. info, critical, success or cautious - defaults to info) */
    context?: Context;
    /** Action array, each containing item text, callback and/or href to action. Max 2 */
    actions?: ActionProps[];
    /** Has a close button, default: true */
    isClosable?: boolean;
    /** A label for the close button that will be used by screenreaders */
    closeButtonLabel?: string;
}

const { block, elem } = bem('Toast', styles);

const ContextIcon = ({ context }: { context: Context }) => {
    switch (context) {
        case 'info':
            return <InfoIcon data-testid="info-icon" />;
        case 'success':
            return <SuccessIcon data-testid="success-icon" />;
        case 'cautious':
            return <CautiousIcon data-testid="cautious-icon" />;
        case 'critical':
            return <CriticalIcon data-testid="critical-icon" />;
        default:
            return assertNever(context, 'Toaster.tsx');
    }
};

export const toast = ({
    title,
    description,
    context = 'info',
    actions,
    isClosable = true,
    closeButtonLabel,
}: ToastProps) =>
    sonnerToast.custom(
        (t) => (
            <div {...block({ context })}>
                <div {...elem('contextIcon')}>
                    <ContextIcon context={context} />
                </div>
                <div {...elem('content')}>
                    <div {...elem('message')}>
                        <h3 className="OneUI-heading-3">{title}</h3>
                        <p className="OneUI-body-text">{description}</p>
                    </div>
                    <div {...elem('actions')}>
                        {actions &&
                            actions.map((action, index) => (
                                <React.Fragment key={`${action.text}`}>
                                    <ActionItem
                                        action={action}
                                        dismissToast={() => sonnerToast.dismiss(t)}
                                    />
                                    {index < actions.length - 1 && <>â€¢</>}
                                </React.Fragment>
                            ))}
                    </div>
                </div>
                {isClosable && (
                    <button
                        {...elem('closeButton')}
                        type="button"
                        onClick={() => sonnerToast.dismiss(t)}
                        aria-label={closeButtonLabel}
                    >
                        <CloseIcon {...elem('closeIcon')} />
                    </button>
                )}
            </div>
        ),
        actions ? { important: true } : {} // focused if actionable
    );

const bemToaster = bem('Toaster', styles);

export const Toaster = ({ duration = 2500, ...props }) => (
    <SonnerToaster
        {...{
            ...props,
            duration,
            closeButton: true,
            offset: 16,
            gap: 8,
            ...bemToaster.block(props),
        }}
    />
);

toast.displayName = 'toast';
Toaster.displayName = 'Toaster';
ContextIcon.displayName = 'ContextIcon';
