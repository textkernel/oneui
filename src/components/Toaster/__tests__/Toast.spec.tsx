import React from 'react';
import '@testing-library/jest-dom';
import { RenderResult, fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Toaster, toast } from '../Toaster';

window.HTMLElement.prototype.setPointerCapture = jest.fn();

describe('<Toast>', () => {
    let view: RenderResult;
    beforeEach(async () => {
        view = render(<Toaster>{}</Toaster>);
    });

    it('should render toast correctly', async () => {
        expect(view.container).toBeEmptyDOMElement();
        toast({
            title: 'Test',
            description: 'Description',
        });
        await waitFor(() => {
            expect(view.container).toMatchSnapshot();
            expect(screen.getByRole('heading', { name: 'Test' })).toBeInTheDocument();
            expect(screen.getByText('Description')).toBeInTheDocument();
        });
    });

    describe('should render the icons correctly with each context', () => {
        it('should render cautious icon', async () => {
            toast({
                description: 'Description',
                context: 'cautious',
            });
            await waitFor(() => {
                expect(screen.getByTestId('cautious-icon')).toBeInTheDocument();
            });
        });

        it('should render info icon', async () => {
            toast({
                description: 'Description',
                context: 'info',
            });
            await waitFor(() => {
                expect(screen.getByTestId('info-icon')).toBeInTheDocument();
            });
        });

        it('should render critical icon', async () => {
            toast({
                description: 'Description',
                context: 'critical',
            });
            await waitFor(() => {
                expect(screen.getByTestId('critical-icon')).toBeInTheDocument();
            });
        });

        it('should render success icon', async () => {
            toast({
                description: 'Description',
                context: 'success',
            });
            await waitFor(() => {
                expect(screen.getByTestId('success-icon')).toBeInTheDocument();
            });
        });
    });

    it('should close automatically after 2500ms', async () => {
        toast({
            title: 'Test',
            description: 'Description',
        });

        await waitFor(() => {
            expect(screen.getByRole('heading', { name: 'Test' })).toBeInTheDocument();
        });

        await waitFor(
            () => {
                expect(view.container).toBeEmptyDOMElement();
            },
            { timeout: 3000 }
        );
    });

    it('should stay open on hover', async () => {
        toast({
            title: 'Test',
            description: 'Description',
        });
        await waitFor(() => {
            expect(screen.getByRole('heading', { name: 'Test' })).toBeInTheDocument();
        });
        const title = screen.getByRole('heading', { name: 'Test' });
        fireEvent.mouseEnter(title);
        await waitFor(
            () => {
                expect(view.container).not.toBeEmptyDOMElement();
                expect(screen.getByRole('heading', { name: 'Test' })).toBeInTheDocument();
            },
            { timeout: 3000 }
        );
    });

    it('should close toast when clicking on the close button', async () => {
        toast({
            description: 'Description',
            closeButtonLabel: 'closeButton',
        });
        await waitFor(() => {
            expect(screen.getByRole('button')).toBeInTheDocument();
        });
        const button = screen.getByLabelText('closeButton');

        const user = userEvent.setup();
        await user.click(button);
        await waitFor(() => {
            expect(view.container).toBeEmptyDOMElement();
        });
    });

    it('should not be focused when not actionable', async () => {
        toast({
            description: 'Description',
        });
        await waitFor(() => {
            expect(screen.getByRole('status')).toHaveAttribute('aria-live', 'polite');
        });
    });

    it('should be focused when actionable', async () => {
        toast({
            description: 'Description',
            actions: [
                {
                    text: 'action1',
                    callback: () => {},
                },
            ],
        });
        await waitFor(() => {
            expect(screen.getByRole('status')).toHaveAttribute('aria-live', 'assertive');
        });
    });

    it('should close when action item callback is clicked', async () => {
        toast({
            description: 'Description',
            actions: [
                {
                    text: 'action1',
                    callback: () => {},
                },
            ],
        });
        await waitFor(() => {
            expect(screen.getByText('action1')).toBeInTheDocument();
        });
        const user = userEvent.setup();
        const actionButton = screen.getByText('action1');
        await user.click(actionButton);
        await waitFor(() => {
            expect(view.container).toBeEmptyDOMElement();
        });
    });

    it('should close when action item href is clicked', async () => {
        toast({
            description: 'Description',
            actions: [{ text: 'action1', href: '/login' }],
        });
        await waitFor(() => {
            expect(screen.getByText('action1')).toBeInTheDocument();
        });
        const user = userEvent.setup();
        const actionButton = screen.getByText('action1');
        await user.click(actionButton);
        await waitFor(() => {
            expect(view.container).toBeEmptyDOMElement();
        });
    });

    it('should close 2500ms after end hover', async () => {
        toast({
            title: 'Test',
            description: 'Description',
        });
        await waitFor(() => {
            expect(screen.getByRole('heading', { name: 'Test' })).toBeInTheDocument();
        });

        const title = screen.getByRole('heading', { name: 'Test' });
        fireEvent.mouseEnter(title);

        await waitFor(
            () => {
                expect(view.container).not.toBeEmptyDOMElement();
                expect(screen.getByRole('heading', { name: 'Test' })).toBeInTheDocument();
            },
            { timeout: 3000 }
        );

        fireEvent.mouseLeave(title);

        await waitFor(
            () => {
                expect(view.container).toBeEmptyDOMElement();
            },
            { timeout: 3000 }
        );
    });
});
