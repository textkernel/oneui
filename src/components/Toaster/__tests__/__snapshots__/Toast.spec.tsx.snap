// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toast> should render toast correctly 1`] = `<div />`;

exports[`<Toast> should render toast correctly 2`] = `
<div>
  <section
    aria-label="Notifications alt+T"
    tabindex="-1"
  >
    <ol
      class="Toaster"
      data-sonner-toaster="true"
      data-theme="light"
      data-x-position="right"
      data-y-position="bottom"
      dir=""
      style="--front-toast-height: 0px; --offset: 16px; --width: 356px; --gap: 8px;"
      tabindex="-1"
    >
      <li
        aria-atomic="true"
        aria-live="polite"
        class=""
        data-dismissible="true"
        data-expanded="false"
        data-front="true"
        data-index="0"
        data-mounted="false"
        data-promise="false"
        data-removed="false"
        data-sonner-toast=""
        data-styled="false"
        data-swipe-out="false"
        data-swiping="false"
        data-visible="true"
        data-x-position="right"
        data-y-position="bottom"
        role="status"
        style="--index: 0; --toasts-before: 0; --z-index: 1; --offset: -8px; --initial-height: 0px;"
        tabindex="0"
      >
        <div
          class="Toast Toast--context_info"
        >
          <div
            class="Toast__contextIcon Toast__contextIcon--context_info"
          >
            <svg
              data-testid="info-icon"
            />
          </div>
          <div
            class="Toast__content"
          >
            <div
              class="Toast__message"
            >
              <h3
                class="OneUI-heading-3"
              >
                Test
              </h3>
              <p
                class="OneUI-body-text"
              >
                Description
              </p>
            </div>
            <div
              class="Toast__actions"
            />
          </div>
          <button
            class="Toast__closeButton"
            type="button"
          >
            <svg
              class="Toast__closeIcon"
              data-testid="default-icon"
            />
          </button>
        </div>
      </li>
    </ol>
  </section>
</div>
`;
